spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/pulsar/PulsarProperties.java

```diff
@@ -943,9 +937,39 @@ public void setTimeout(Duration timeout) {
 			this.timeout = timeout;
 		}
 
-		void validate(String prefix) {
-			Assert.state((this.enabled || !this.required), "If transactions are required "
-					+ "they must also be enabled - consult your '%s.transaction' properties.".formatted(prefix));
+		void validate() {
+			if (this.required && !this.enabled) {
+				String requiredProp = "%s.required".formatted(this.propertyPath());
+				String enabledProp = "%s.enabled".formatted(this.propertyPath());
+				throw new InvalidConfigurationPropertyValueException(requiredProp, this.required,
+						"Transactions must be enabled in order to be required. "
+								+ "Either set %s to 'true' or make transactions optional by setting %s to 'false'"
+									.formatted(enabledProp, requiredProp));
+			}
+		}
+
+		/**
+		 * Gets the property path that the transaction properties are mapped to.
+		 * @return the property path that the transaction properties are mapped to
+		 */
+		protected abstract String propertyPath();
+
+	}
+
+	static class TemplateTransaction extends Transaction {
+
+		@Override
+		protected String propertyPath() {
+			return "spring.pulsar.template.transaction";
```

Comment:
This avoids callers needing to know internals about where the property is mapped.

Reply:
The `protected` method will be available to users who sub-class `PulsarProperties`. I would prefer that we keep the surface area of the API as small as possible so I've reworked this and it's now all encapsulated in the mapper.